File: backend/app/pattern_expander.py
"""
Pattern Expander for XSG

Handles template inheritance (extends) and parameter expansion ({{paramName}})
Similar to frontend/src/lib/paramExpander.ts
"""

import re
from copy import deepcopy
from pathlib import Path
from typing import Any, Dict, Optional

import yaml

from .models import ParamDef, XSGPattern


class PatternExpander:
    """Pattern expander with extends resolution and parameter substitution"""

    def __init__(self, patterns_dir: str = None):
        """
        Initialize pattern expander

        Args:
            patterns_dir: Directory containing pattern files (defaults to project_root/patterns)
        """
        if patterns_dir is None:
            # Default to project_root/patterns
            project_root = Path(__file__).parent.parent.parent
            patterns_dir = str(project_root / "patterns")
        self.patterns_dir = Path(patterns_dir)

    def expand(
        self, pattern: XSGPattern, user_params: Optional[Dict[str, Any]] = None
    ) -> XSGPattern:
        """
        Expand pattern with parameter substitution

        Args:
            pattern: Pattern definition with {{paramName}} variables
            user_params: User-provided parameter values (from URL query or API)

        Returns:
            Pattern with expanded parameters
        """
        user_params = user_params or {}

        # Merge default params with user params
        params = self._resolve_params(pattern.params or {}, user_params)

        # Deep clone and expand
        pattern_dict = pattern.model_dump(mode="json", exclude_none=True)
        expanded_dict = self._expand_object(pattern_dict, params)

        # Return as XSGPattern
        return XSGPattern(**expanded_dict)

    def resolve_extends(self, pattern: XSGPattern) -> XSGPattern:
        """
        Resolve template inheritance (extends)

        Args:
            pattern: Pattern that may have extends property

        Returns:
            Resolved pattern with base pattern merged
        """
        if not pattern.extends:
            return pattern

        # Load base pattern
        base_pattern = self._load_pattern_file(pattern.extends)

        # Recursively resolve base pattern's extends
        resolved_base = self.resolve_extends(base_pattern)

        # Merge base and child patterns
        merged = self._merge_patterns(resolved_base, pattern)

        return merged

    def _resolve_params(
        self, param_defs: Dict[str, ParamDef], user_params: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Resolve parameter values (default + user overrides)

        Args:
            param_defs: Parameter definitions
            user_params: User-provided values

        Returns:
            Resolved parameter values
        """
        resolved = {}

        for key, param_def in param_defs.items():
            if key in user_params:
                # User provided value - coerce to correct type
                resolved[key] = self._coerce_type(user_params[key], param_def.type.value)
            elif param_def.default is not None:
                # Default value
                resolved[key] = param_def.default

        return resolved

    def _coerce_type(self, value: Any, param_type: str) -> Any:
        """
        Coerce value to the specified type

        Args:
            value: Value to coerce
            param_type: Target type ('number', 'string', 'color', 'boolean')

        Returns:
            Coerced value
        """
        if param_type == "number":
            return float(value) if "." in str(value) else int(value)
        elif param_type == "boolean":
            if isinstance(value, bool):
                return value
            return str(value).lower() == "true"
        elif param_type in ("string", "color"):
            return str(value)
        else:
            return value

    def _expand_object(self, obj: Any, params: Dict[str, Any]) -> Any:
        """
        Recursively expand {{paramName}} in all string values

        Args:
            obj: Object to expand
            params: Parameter values

        Returns:
            Expanded object
        """
        if isinstance(obj, str):
            return self._expand_string(obj, params)
        elif isinstance(obj, dict):
            return {key: self._expand_object(value, params) for key, value in obj.items()}
        elif isinstance(obj, list):
            return [self._expand_object(item, params) for item in obj]
        else:
            return obj

    def _expand_string(self, s: str, params: Dict[str, Any]) -> Any:
        """
        Expand {{paramName}} in a single string

        Args:
            s: String to expand
            params: Parameter values

        Returns:
            Expanded value (may be string or original type if single variable)
        """
        # Check if the entire string is a single variable reference
        single_var_match = re.match(r"^{{(\w+)}}$", s)
        if single_var_match:
            param_name = single_var_match.group(1)
            if param_name in params:
                return params[param_name]
            return s

        # Replace all {{paramName}} occurrences
        def replace_var(match):
            param_name = match.group(1)
            if param_name in params:
                return str(params[param_name])
            return match.group(0)  # Keep original if param not found

        return re.sub(r"{{(\w+)}}", replace_var, s)

    def _merge_patterns(self, base: XSGPattern, child: XSGPattern) -> XSGPattern:
        """
        Merge base and child patterns

        Args:
            base: Base pattern (from extends)
            child: Child pattern

        Returns:
            Merged pattern
        """
        # Merge params: individual property merging (child overrides base)
        merged_params = {}
        if base.params:
            merged_params = deepcopy(base.params)

        if child.params:
            for key, child_param in child.params.items():
                if key in merged_params:
                    # Merge child param properties with base param
                    base_param_dict = merged_params[key].model_dump(exclude_none=True)
                    child_param_dict = child_param.model_dump(exclude_none=True)
                    base_param_dict.update(child_param_dict)
                    merged_params[key] = ParamDef(**base_param_dict)
                else:
                    # New param from child
                    merged_params[key] = child_param

        # Build merged pattern
        merged_dict = {}

        # Canvas: child overrides base
        if child.canvas is not None:
            merged_dict["canvas"] = child.canvas
        elif base.canvas is not None:
            merged_dict["canvas"] = base.canvas

        # Params: merged
        if merged_params:
            merged_dict["params"] = merged_params

        # Nodes: child overrides base (or use base if child has none)
        if child.nodes is not None:
            merged_dict["nodes"] = child.nodes
        elif base.nodes is not None:
            merged_dict["nodes"] = base.nodes

        return XSGPattern(**merged_dict)

    def _load_pattern_file(self, path: str) -> XSGPattern:
        """
        Load a pattern file by path

        Args:
            path: Pattern file path (relative to patterns/ directory)

        Returns:
            Loaded pattern
        """
        # Normalize path
        if path.startswith("/"):
            file_path = Path(path[1:])
        else:
            file_path = self.patterns_dir / path

        # Read YAML
        with open(file_path, "r", encoding="utf-8") as f:
            data = yaml.safe_load(f)

        return XSGPattern(**data)

    def _resolve_extends_dict(self, pattern_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Resolve extends using dictionary data (no Pydantic validation)

        Args:
            pattern_data: Pattern data as dict

        Returns:
            Resolved pattern data as dict
        """
        if "extends" not in pattern_data or pattern_data["extends"] is None:
            return pattern_data

        # Load base pattern
        extends_path = pattern_data["extends"]
        if extends_path.startswith("/"):
            file_path = Path(extends_path[1:])
        else:
            file_path = self.patterns_dir / extends_path

        with open(file_path, "r", encoding="utf-8") as f:
            base_data = yaml.safe_load(f)

        # Recursively resolve base pattern's extends
        resolved_base = self._resolve_extends_dict(base_data)

        # Merge patterns (dict-based)
        merged = self._merge_pattern_dicts(resolved_base, pattern_data)

        return merged

    def _merge_pattern_dicts(
        self, base: Dict[str, Any], child: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Merge base and child pattern dicts

        Args:
            base: Base pattern dict
            child: Child pattern dict

        Returns:
            Merged pattern dict
        """
        merged = {}

        # Canvas: child overrides base
        if "canvas" in child:
            merged["canvas"] = child["canvas"]
        elif "canvas" in base:
            merged["canvas"] = base["canvas"]

        # Params: merge individual parameters
        merged_params = {}
        if "params" in base:
            merged_params = deepcopy(base["params"])

        if "params" in child:
            for key, child_param in child["params"].items():
                if key in merged_params:
                    # Merge child param properties with base param
                    merged_params[key] = {**merged_params[key], **child_param}
                else:
                    merged_params[key] = child_param

        if merged_params:
            merged["params"] = merged_params

        # Nodes: child overrides base
        if "nodes" in child:
            merged["nodes"] = child["nodes"]
        elif "nodes" in base:
            merged["nodes"] = base["nodes"]

        return merged


# Global expander instance
_default_expander: Optional[PatternExpander] = None


def get_pattern_expander(patterns_dir: str = None) -> PatternExpander:
    """
    Get or create default pattern expander

    Args:
        patterns_dir: Directory containing pattern files (defaults to project_root/patterns)

    Returns:
        PatternExpander instance
    """
    global _default_expander

    if _default_expander is None:
        _default_expander = PatternExpander(patterns_dir=patterns_dir)

    return _default_expander


def expand_pattern(
    pattern: XSGPattern, user_params: Optional[Dict[str, Any]] = None
) -> XSGPattern:
    """
    Convenience function to expand a pattern

    Args:
        pattern: Pattern to expand
        user_params: User-provided parameter values

    Returns:
        Expanded pattern
    """
    expander = get_pattern_expander()

    # First resolve extends
    resolved = expander.resolve_extends(pattern)

    # Then expand parameters
    expanded = expander.expand(resolved, user_params)

    return expanded


def expand_pattern_dict(
    pattern_data: Dict[str, Any], user_params: Optional[Dict[str, Any]] = None
) -> Dict[str, Any]:
    """
    Expand pattern from dictionary (without Pydantic validation)

    This allows processing patterns with {{template}} variables before validation.

    Args:
        pattern_data: Pattern data as dictionary
        user_params: User-provided parameter values

    Returns:
        Expanded pattern as dictionary
    """
    expander = get_pattern_expander()

    # 1. Resolve extends (recursive, dict-based)
    resolved_data = expander._resolve_extends_dict(pattern_data)

    # 2. Resolve parameters
    param_defs = resolved_data.get("params", {})
    user_params = user_params or {}

    # Convert ParamDef dicts to resolved values
    params = {}
    for key, param_def_data in param_defs.items():
        if key in user_params:
            params[key] = expander._coerce_type(
                user_params[key], param_def_data.get("type", "string")
            )
        elif "default" in param_def_data:
            params[key] = param_def_data["default"]

    # 3. Expand {{vars}} in the data
    expanded_data = expander._expand_object(resolved_data, params)

    return expanded_data

File: frontend/src-tauri/src/pattern_expander.rs
use anyhow::{Context, Result};
use serde_json::Value;
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

/// Pattern expander for handling template inheritance (extends) and parameter expansion
pub struct PatternExpander {
    patterns_dir: PathBuf,
}

impl PatternExpander {
    /// Create a new PatternExpander with the specified patterns directory
    pub fn new(patterns_dir: PathBuf) -> Self {
        Self { patterns_dir }
    }

    /// Expand a pattern with parameter substitution
    ///
    /// # Arguments
    /// * `pattern` - Pattern data with {{paramName}} variables
    /// * `user_params` - User-provided parameter values (from URL query or API)
    pub fn expand(&self, pattern: &mut Value, user_params: &HashMap<String, String>) -> Result<()> {
        // Resolve parameter values
        let params = self.resolve_params(pattern, user_params)?;

        // Expand {{paramName}} in all string values
        self.expand_object(pattern, &params);

        Ok(())
    }

    /// Resolve template inheritance (extends)
    ///
    /// # Arguments
    /// * `pattern` - Pattern that may have extends property
    pub fn resolve_extends(&self, pattern: &mut Value) -> Result<()> {
        // Check if pattern has extends property
        if !pattern.is_object() {
            return Ok(());
        }

        let extends_path = pattern
            .get("extends")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string());

        if extends_path.is_none() {
            return Ok(());
        }

        let extends_path = extends_path.unwrap();

        // Load base pattern
        let mut base_pattern = self.load_pattern_file(&extends_path)?;

        // Recursively resolve base pattern's extends
        self.resolve_extends(&mut base_pattern)?;

        // Merge base and child patterns
        self.merge_patterns(&mut base_pattern, pattern);

        // Replace current pattern with merged result
        *pattern = base_pattern;

        Ok(())
    }

    /// Resolve parameter values (default + user overrides)
    fn resolve_params(
        &self,
        pattern: &Value,
        user_params: &HashMap<String, String>,
    ) -> Result<HashMap<String, Value>> {
        let mut resolved = HashMap::new();

        // Get param definitions from pattern
        if let Some(params_obj) = pattern.get("params").and_then(|v| v.as_object()) {
            for (key, param_def) in params_obj {
                if let Some(user_value) = user_params.get(key) {
                    // User provided value - coerce to correct type
                    let param_type = param_def
                        .get("type")
                        .and_then(|v| v.as_str())
                        .unwrap_or("string");
                    resolved.insert(key.clone(), self.coerce_type(user_value, param_type)?);
                } else if let Some(default_value) = param_def.get("default") {
                    // Default value
                    resolved.insert(key.clone(), default_value.clone());
                }
            }
        }

        Ok(resolved)
    }

    /// Coerce value to the specified type
    fn coerce_type(&self, value: &str, param_type: &str) -> Result<Value> {
        match param_type {
            "number" => {
                // Try to parse as integer first, then float
                if let Ok(int_val) = value.parse::<i64>() {
                    Ok(Value::Number(int_val.into()))
                } else if let Ok(float_val) = value.parse::<f64>() {
                    Ok(Value::Number(
                        serde_json::Number::from_f64(float_val)
                            .context("Invalid float value")?,
                    ))
                } else {
                    Err(anyhow::anyhow!("Cannot parse '{}' as number", value))
                }
            }
            "boolean" => {
                let bool_val = value.to_lowercase() == "true";
                Ok(Value::Bool(bool_val))
            }
            "string" | "color" => Ok(Value::String(value.to_string())),
            _ => Ok(Value::String(value.to_string())),
        }
    }

    /// Recursively expand {{paramName}} in all string values
    fn expand_object(&self, obj: &mut Value, params: &HashMap<String, Value>) {
        match obj {
            Value::String(s) => {
                *obj = self.expand_string(s, params);
            }
            Value::Object(map) => {
                for (_, value) in map.iter_mut() {
                    self.expand_object(value, params);
                }
            }
            Value::Array(arr) => {
                for item in arr.iter_mut() {
                    self.expand_object(item, params);
                }
            }
            _ => {}
        }
    }

    /// Expand {{paramName}} in a single string
    fn expand_string(&self, s: &str, params: &HashMap<String, Value>) -> Value {
        // Check if the entire string is a single variable reference
        if let Some(var_name) = self.extract_single_var(s) {
            if let Some(value) = params.get(var_name) {
                return value.clone();
            }
            return Value::String(s.to_string());
        }

        // Replace all {{paramName}} occurrences
        let mut result = s.to_string();
        for (key, value) in params {
            let placeholder = format!("{{{{{}}}}}", key);
            if result.contains(&placeholder) {
                let replacement = match value {
                    Value::String(s) => s.clone(),
                    Value::Number(n) => n.to_string(),
                    Value::Bool(b) => b.to_string(),
                    _ => value.to_string(),
                };
                result = result.replace(&placeholder, &replacement);
            }
        }

        Value::String(result)
    }

    /// Extract variable name if string is a single {{varName}} reference
    fn extract_single_var<'a>(&self, s: &'a str) -> Option<&'a str> {
        if s.starts_with("{{") && s.ends_with("}}") && s.matches("{{").count() == 1 {
            Some(&s[2..s.len() - 2])
        } else {
            None
        }
    }

    /// Merge base and child patterns
    fn merge_patterns(&self, base: &mut Value, child: &Value) {
        if !base.is_object() || !child.is_object() {
            return;
        }

        let base_obj = base.as_object_mut().unwrap();
        let child_obj = child.as_object().unwrap();

        // Merge params: individual property merging (child overrides base)
        if let Some(child_params) = child_obj.get("params") {
            if let Some(child_params_obj) = child_params.as_object() {
                if let Some(base_params) = base_obj.get_mut("params") {
                    if let Some(base_params_obj) = base_params.as_object_mut() {
                        // Merge each parameter
                        for (key, child_param) in child_params_obj {
                            if let Some(base_param) = base_params_obj.get_mut(key) {
                                // Merge child param properties with base param
                                self.merge_param_properties(base_param, child_param);
                            } else {
                                // New param from child
                                base_params_obj.insert(key.clone(), child_param.clone());
                            }
                        }
                    }
                } else {
                    // No base params, use child params
                    base_obj.insert("params".to_string(), child_params.clone());
                }
            }
        }

        // Canvas: child overrides base
        if let Some(child_canvas) = child_obj.get("canvas") {
            base_obj.insert("canvas".to_string(), child_canvas.clone());
        }

        // Nodes: child overrides base
        if let Some(child_nodes) = child_obj.get("nodes") {
            base_obj.insert("nodes".to_string(), child_nodes.clone());
        }

        // Name and category: child overrides base
        if let Some(child_name) = child_obj.get("name") {
            base_obj.insert("name".to_string(), child_name.clone());
        }
        if let Some(child_category) = child_obj.get("category") {
            base_obj.insert("category".to_string(), child_category.clone());
        }

        // Remove extends from merged pattern
        base_obj.remove("extends");
    }

    /// Merge child param properties with base param
    fn merge_param_properties(&self, base_param: &mut Value, child_param: &Value) {
        if !base_param.is_object() || !child_param.is_object() {
            return;
        }

        let base_param_obj = base_param.as_object_mut().unwrap();
        let child_param_obj = child_param.as_object().unwrap();

        // Child properties override base properties
        for (key, value) in child_param_obj {
            base_param_obj.insert(key.clone(), value.clone());
        }
    }

    /// Load a pattern file by path
    fn load_pattern_file(&self, path: &str) -> Result<Value> {
        // Normalize path
        let file_path = if path.starts_with('/') {
            PathBuf::from(&path[1..])
        } else {
            self.patterns_dir.join(path)
        };

        // Read YAML
        let content = fs::read_to_string(&file_path)
            .with_context(|| format!("Failed to read pattern file: {}", file_path.display()))?;

        let yaml: Value = serde_yaml::from_str(&content)
            .with_context(|| format!("Failed to parse YAML: {}", file_path.display()))?;

        Ok(yaml)
    }
}

/// Get the patterns directory path
pub fn get_patterns_dir() -> Result<PathBuf> {
    // In development: use project_root/patterns
    // In production: use exe_dir/patterns or resources/patterns

    if cfg!(debug_assertions) {
        // Development mode
        let current_dir = std::env::current_dir()?;

        // Try to find project root by looking for patterns directory
        let mut search_dir = current_dir.clone();
        loop {
            let patterns_dir = search_dir.join("patterns");
            if patterns_dir.exists() && patterns_dir.is_dir() {
                return Ok(patterns_dir);
            }

            if !search_dir.pop() {
                break;
            }
        }

        // Fallback: use current_dir/patterns
        Ok(current_dir.join("patterns"))
    } else {
        // Production mode
        let exe_path = std::env::current_exe()?;
        let exe_dir = exe_path.parent().context("Cannot get exe directory")?;
        Ok(exe_dir.join("patterns"))
    }
}
